R/validate-pds-trips.R

```diff
@@ -131,3 +205,97 @@ get_preprocessed_trips <- function(pars){
                       options = pars$storage$google$options)
   readr::read_rds(file = pds_trips_rds)
 }
+
+
+#' Merge short intervalled trips
+#'
+#' PDS data are quite patchy, sometimes an actual single trip is recorded as 2
+#' different trips. This function calculates the temporal and the geographical
+#' distance between two consecutive trips for each boat, merging those consecutive
+#' trips that start before 180 minutes from the following within a range of 2 km.
+#'
+#' @param x A data frame containing raw pds trips data.
+#' @param consecutive_time Time between consecutive trips (hrs).
+#' @param consecutive_distance Distance between consecutive trips (Km).
+#'
+#' @return A dataframe with pds trips including a new column `associated_to`
+#' indicating the consecutive trips to be merged.
+#'
+#' @importFrom rlang .data
+#' @export
+#'
+
+merge_consecutive_trips <- function(x,
+                                    consecutive_time = NULL,
+                                    consecutive_distance = NULL) {
+  x %>%
+    dplyr::mutate(
+      end_start_distance = get_distance(.),
+      end_start_distance = dplyr::case_when(
+        is.na(.data$end_start_time) ~ NA_real_,
+        TRUE ~ .data$end_start_distance
+      )
+    ) %>%
+    dplyr::group_by(.data$Boat) %>%
+    dplyr::mutate(start_end_diff = difftime(dplyr::lead(.data$Started),
+                                            .data$Ended,
+                                            units = "mins"
+    )) %>%
+    dplyr::mutate(
+      associated_to = dplyr::case_when(
+        .data$end_start_time < consecutive_time * 60 &
+          .data$end_start_distance < consecutive_distance * 1000 ~ .data$Trip,
+        TRUE ~ NA_integer_
+      ),
+      associated_to = dplyr::lag(.data$associated_to),
+      associated_to = dplyr::coalesce(.data$associated_to, .data$Trip)
+    ) %>%
+    dplyr::ungroup() %>%
+    dplyr::select(.data$Trip:.data$timetrace_dispersion, .data$associated_to) %>%
+    dplyr::group_by(.data$associated_to) %>%
+    dplyr::summarise(
+      Trip = dplyr::first(.data$Trip),
+      Started = dplyr::first(.data$Started),
+      Ended = dplyr::last(.data$Ended),
+      Boat = dplyr::first(.data$Boat),
+      `Boat Name` = dplyr::first(.data$`Boat Name`),
+      `Boat Gear` = dplyr::first(.data$`Boat Gear`),
+      Community = dplyr::first(.data$Community),
+      `Duration (Seconds)` = sum(.data$`Duration (Seconds)`),
+      `Range (Meters)` = sum(.data$`Range (Meters)`),
+      `Distance (Meters)` = sum(.data$`Distance (Meters)`),
+      IMEI = dplyr::first(.data$IMEI),
+      `Device Id` = dplyr::first(.data$`Device Id`),
+      `Last Seen` = dplyr::last(.data$`Last Seen`),
+      start_end_distance = sum(.data$start_end_distance),
```

Comment:
It probably makes no practical difference, but shouldn't this distance be recalculated for the ones that were merged?